sdks/java/io/parquet/src/main/java/org/apache/beam/sdk/io/parquet/ParquetIO.java

```diff
@@ -235,12 +283,164 @@ public ReadFiles withAvroDataModel(GenericData model) {
+        for (int i = 0; i < currentBlock; i++) {
+          reader.skipNextRowGroup();
+        }
+        while (tracker.tryClaim(currentBlock)) {
+          PageReadStore pages = reader.readNextRowGroup();
+          currentBlock += 1;
+          RecordReader<GenericRecord> recordReader =
+              columnIO.getRecordReader(
+                  pages, recordConverter, filterRecords ? filter : FilterCompat.NOOP);
+          GenericRecord read;
+          long currentRow = 0;
+          long totalRows = pages.getRowCount();
+          while (currentRow < totalRows) {
+            outputReceiver.output(recordReader.read());
+            currentRow += 1;
+          }
+        }
+      }
+
+      private Configuration setConf() throws Exception {
+        Configuration conf = new Configuration();
+        GenericData model = null;
+        if (modelClass != null) {
+          model = (GenericData) modelClass.getMethod("get").invoke(null);
+        }
+        if (model != null
+            && (model.getClass() == GenericData.class || model.getClass() == SpecificData.class)) {
+          conf.setBoolean(AvroReadSupport.AVRO_COMPATIBILITY, true);
+        } else {
+          conf.setBoolean(AvroReadSupport.AVRO_COMPATIBILITY, false);
+        }
+        return conf;
+      }
+
+      @GetInitialRestriction
+      public OffsetRange getInitialRestriction(@Element FileIO.ReadableFile file) throws Exception {
+        InputFile inputFile = getInputFile(file);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        return new OffsetRange(0, reader.getRowGroups().size());
+      }
+
+      @SplitRestriction
+      public void split(@Restriction OffsetRange restriction, OutputReceiver<OffsetRange> out) {
+        for (OffsetRange range : restriction.split(1, 0)) {
+          out.output(range);
+        }
+      }
+
+      @NewTracker
+      public OffsetRangeTracker newTracker(@Restriction OffsetRange restriction) {
+        return new OffsetRangeTracker(restriction);
+      }
+
+      @GetSize
+      public double getSize(@Element FileIO.ReadableFile file, @Restriction OffsetRange restriction)
+          throws Exception {
+        InputFile inputFile = getInputFile(file);
+        Configuration conf = setConf();
+        ParquetReadOptions options = HadoopReadOptions.builder(conf).build();
+        ParquetFileReader reader = ParquetFileReader.open(inputFile, options);
+        if (restriction == null) {
+          return reader.getRecordCount();
+        } else {
+          long start = restriction.getFrom();
+          long end = restriction.getTo();
+          List<BlockMetaData> blocks = reader.getRowGroups();
+          double size = 0;
+          for (long i = start; i < end; i++) {
+            size += blocks.get((int) i).getRowCount();
+          }
+          return size;
```

Comment:
If it's feasible, it would be nice to track an estimated avg row size like `KafkaIO`. Then the size will be avgSize * rowCount. @chamikaramj Do you think it's necessary to do so?

Reply:
I think get size is used to get an idea of the current process and for the purpose of optimized split. Since the size of each row is similar for a given file, the number of row is a fairly accurate representation of the work, multiply with a constant does not change too much when we compare. The previous methods of printing the progress also used number of rows(records) in their progress report.

Reply:
If you use byte sizes for tracking implementation here will end up being trivial.